### PR TITLE
SystemUIs authorization

### PR DESCRIPTION
**Pull request by tomusdrw**
It's now required for SystemUIs (SignerUIs) to authorize when connecting to WebSockets.
1. Parity can generate tokens stored in `~/.parity/signer/authcodes` using `parity signer new-token` command.
2. User types the code into SignerUI (popup dialog should be displayed)
3. SignerUI should store the code somewhere (localStorage for now) for future re-use.

How it works now:
1. Tokens are stored in plain text in `~/.parity/signer/authcodes`
2. SystemUI sends: `sha3(<code>:<timestamp>), <timestamp>)` as authorization
3. Parity checks if timestamp is recent (2 seconds) and hash matches.

Problems:
1. User has to type the token once when using built-in minimal system-ui. 
2. Installer has to generate and provide the token to chrome-extension (if possible) - otherwise user would have to paste the token.
3. If browsing in private mode user will have to provide tokens each time.

Possible improvements:
1. Locking the token (for 2 seconds) when used (to avoid replaying the same hash+time)
2. ???

Let me know if you see a better way to handle this.
#1175

### CODE REVIEW

File: signer/src/authcode_store.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			codes: codes,
+			now: now,
+		}
+	}
+
+	/// Checks if given hash is correct identifier of `SignerUI`
+	pub fn is_valid(&self, hash: &H256, time: u64) -> bool {
+		let now = self.now.now();
+		// check time
+		if time >= now + TIME_THRESHOLD || time <= now - TIME_THRESHOLD {
```
Comment (debris) [inline signer/src/authcode_store.rs]: 5 secs may be not enough :)

File: signer/src/authcode_store.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	pub fn generate_new(&mut self) -> io::Result<String> {
+		let mut rng = try!(OsRng::new());
+		let code = rng.gen_ascii_chars().take(TOKEN_LENGTH).collect::<String>();
+		let readable_code = code.as_bytes()
+			.chunks(4)
+			.filter_map(|f| String::from_utf8(f.to_vec()).ok())
+			.collect::<Vec<String>>()
+			.join("-");
+		info!(target: "signer", "New authentication token generated.");
+		self.codes.push(code);
```
Comment (debris) [inline signer/src/authcode_store.rs]: `self.codes` grows infinitely. Can it cause an issue?

**Comment by debris** [comment]
> 1. Locking the token (for 2 seconds) when used (to avoid replaying the same hash+time)

Maybe SystemUI should send something like `sha3(<code>:<timestamp>;<index>), <timestamp>, <index>)`


where `index` may be an integer
- representing number of requests in current second
- random number 

authcode_store would have to store these integer in memory, but not in a file, cause it is reset within a second

File: signer/src/authcode_store.rs (hunk 2 above)
Comment (tomusdrw) [inline signer/src/authcode_store.rs]: It's only invoked when `parity signer new-token` is run by user from CLI. I think it's not an issue.

**Comment by tomusdrw** [comment]
This authorization token is not used for each request. It's only used to connect to the server.
Single code per request is still something to consider in future.

The idea is to avoid "untrusted" SystemUIs to connect to full set of APIs (you can connect to websockets without any CORS required).

**Comment by debris** [comment]
Then locking the token should be enough :) LGTM

**Pull request merged**